server/src/main/java/org/elasticsearch/action/bulk/BulkShardRequestInferenceProvider.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.action.bulk;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.DocWriteRequest;
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.action.support.RefCountingRunnable;
+import org.elasticsearch.action.update.UpdateRequest;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.common.TriConsumer;
+import org.elasticsearch.core.Releasable;
+import org.elasticsearch.inference.InferenceProvider;
+import org.elasticsearch.inference.InferenceResults;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+public class BulkShardRequestInferenceProvider {
```

Comment:
Extracted inference related code into its own class. @benwtrent , WDYT about the current direction?

Reply:
Yeah, this is more like it. I think the failure handling and everything should happen here. Meaning the `onBulkItemInferenceFailure` should be handled here.

I think this along with better encapsulating the InferenceProvider work that removes the need for `ROOT_RESULT_FIELD` etc. will clean this up nicely.

Reply:
I changed the dependency, so this class defines the constants and they are used from the field mappers. LMK if this addresses your concerns.